core/cc/archive.cpp

```diff
@@ -40,11 +49,147 @@ void must_truncate(int fd, off_t length) {
 }  // anonymous namespace
 
 namespace core {
+#if defined GAPID_ARCHIVE_USE_MMAP
+Archive::RecordFile::RecordFile()
+    : fd(-1), base(nullptr), end(0), capacity(0) {}
+
+bool Archive::RecordFile::open(const std::string &filename) {
+  fd = ::open(filename.c_str(), O_RDWR | O_CREAT, S_IRWXU);
+  if (fd == -1) {
+    return false;
+  }
+  struct stat st;
+  ::fstat(fd, &st);
+  end = st.st_size;
+  capacity = end;
+  return map();
+}
+
+void Archive::RecordFile::close() {
+  if (fd == -1) return;
+  if (!unmap()) {
+    GAPID_FATAL("Unable to unmap archive file.");
+  }
+  // Set the proper file size when we close the file.
+  if (::ftruncate(fd, end)) {
+    GAPID_FATAL("Unable to truncate archive file.");
+  }
+  ::close(fd);
+}
+
+bool Archive::RecordFile::read(uint64_t offset, void *buf, size_t size) {
+  if (offset + size > end) {
+    return false;
+  }
+  memcpy(buf, at(offset), size);
+  return true;
+}
+
+bool Archive::RecordFile::append(const void *buf, size_t size) {
+  if (!reserve(end + size)) {
+    return false;
+  }
+  memcpy(at(end), buf, size);
+  end += size;
+  return true;
+}
+
+uint64_t Archive::RecordFile::size() { return end; }
+
+bool Archive::RecordFile::resize(uint64_t size) {
+  if (size > capacity) {
+    if (!reserve(size)) {
+      return false;
+    }
+  }
+  end = size;
+  return true;
+}
+
+bool Archive::RecordFile::reserve(uint64_t requiredCapacity) {
+  if (requiredCapacity <= capacity) {
+    return true;
+  }
+
+  if (!unmap()) {
+    GAPID_FATAL("Unable to unmap archive file.");
+  }
+
+  // Reserve at least 1.5 time the size.
+  if (requiredCapacity < end * 3 / 2) {
+    requiredCapacity = end * 3 / 2;
+  }
 
-Archive::Archive(const std::string& archiveName) {
+  // Align to the next 4k boundary, not necessary but good to have.
+  requiredCapacity = (requiredCapacity + 0xfff) & ~(uint64_t)0xfff;
+
+  if (::ftruncate(fd, requiredCapacity)) {
+    GAPID_FATAL("Unable to ftruncate(grow) archive file.");
+  }
+
+  capacity = requiredCapacity;
+
+  if (!map()) {
+    GAPID_FATAL("Unable to map archive file.");
+  }
+
+  return true;
+}
+
+bool Archive::RecordFile::unmap() {
+  if (!base) return true;
+  if (::munmap(base, capacity)) {
+    return false;
+  }
+  base = nullptr;
+  return true;
+}
+
+bool Archive::RecordFile::map() {
+  if (base || capacity == 0) {
+    return true;  // Already mapped or don't need to map.
+  }
+  base = ::mmap(nullptr, capacity, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
+  return base != nullptr;
+}
+
+#else
```

Comment:
This is a large preprocessor block. Please add a comment describing what the condition is.